### PR TITLE
Created new login-link interface.

### PR DESCRIPTION
**Pull request by jonahbron**
Created an awu-login-link component to prompt the user for an interface.
Fixed the login-link worker to run in the worker environment properly
(jsonwebtoken doesn't work).  Wired up and verified to work correctly.

image

image

### CODE REVIEW

File: frontend/src/element/login-link.ts [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+  form!: HTMLFormElement;
+
+  @state()
+  isLoading = false;
+  @state()
+  isComplete = false;
+
+  render(): TemplateResult {
+    return html`
+      <div class="completed ${classMap({ 'not-completed': !this.isComplete })}">
```
Comment (taralx) [inline frontend/src/element/login-link.ts:36]: So this has class "completed not-completed" at the start?

File: frontend/src/element/login-link.ts [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+        <h2>Check your inbox</h2>
+        <p>
+          If that email is associated with a membership, we will send you an
+          email with a login link.
+        </p>
+        <p>
+          If you don't get a link, make sure you entered the correct email
```
Comment (taralx) [inline frontend/src/element/login-link.ts:43]: Always a good idea to recommend people check their spam box and also to tell them from what address the email comes.

File: frontend/src/element/login-link.ts [hunk 3]
```diff
@@ -0,0 +1,119 @@
...
+      </form>
+    `;
+  }
+
+  async submit(event: Event): Promise<void> {
+    event.preventDefault();
+
+    this.isLoading = true;
+    const body = new FormData(this.form);
+    const uid = crypto.randomUUID();
```
Comment (taralx) [inline frontend/src/element/login-link.ts:82]: What's this for?

File: frontend/src/element/login-link.ts [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+      const result = await fetch(window.LOGIN_LINK_TOKEN_API, {
+        method: 'POST',
+        body,
+      });
+
+      if (result.ok) {
+        // TODO: save uid to cookie
+        this.isComplete = true;
+      } else {
+        const { error } = await result.json();
```
Comment (taralx) [inline frontend/src/element/login-link.ts:94]: This can and will throw on things that are not JSON, like network errors.

File: frontend/src/element/login-link.ts [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+        console.error(error);
+        throw error;
```
Comment (taralx) [inline frontend/src/element/login-link.ts:96]: Please don't throw arbitrary objects. It's a strongly discouraged practice.

File: frontend/src/element/login-link.ts [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+      }
+      this.isComplete = true;
+    } catch (e) {
+      if (e) {
+        console.error(e);
+        alert(e);
```
Comment (taralx) [inline frontend/src/element/login-link.ts:102]: Is this the final form for error handling?

File: frontend/src/element/login-link.ts [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+    body.set('uid', uid);
+    try {
+      const result = await fetch(window.LOGIN_LINK_TOKEN_API, {
+        method: 'POST',
+        body,
+      });
+
+      if (result.ok) {
+        // TODO: save uid to cookie
+        this.isComplete = true;
```
Comment (taralx) [inline frontend/src/element/login-link.ts:92]: Either this or line 98 is redundant.

File: frontend/src/element/login-link.ts [hunk 8]
```diff
@@ -0,0 +1,119 @@
...
+    }
+  }
+}
+
+declare global {
+  interface Window {
+    LOGIN_LINK_TOKEN_API: string;
+  }
+
+  // TODO: update Typescript version to get randomUUID in the type def.
```
Comment (taralx) [inline frontend/src/element/login-link.ts:115]: Install @types/dom and it will happen automatically IIRC.

File: frontend/src/element/signup.ts [hunk 9]
```diff
@@ -682,7 +682,7 @@ export class Signup extends LitElement {
               ${repeat(
                 this.availableRegions,
                 (regionData) => regionData[1],
-                (regionData, index) => html`
```
Comment (taralx) [inline frontend/src/element/signup.ts:685]: This appears to be an unrelated change.

File: login-link-token-worker/wrangler.toml [hunk 10]
```diff
@@ -1,14 +1,12 @@
...
+compatibility_date = "2022-06-18"
 
 [dev]
-port = 8789
+port = 8790
 
 [env.production]
-vars = { PLAID_ORIGIN = "https://production.plaid.com" }
\ No newline at end of file
+vars = { PLAID_ORIGIN = "https://production.plaid.com" }
```
Comment (taralx) [inline login-link-token-worker/wrangler.toml:12]: Is plaid handling our login functions?

**Comment by jonahbron** [comment]
Obsolete. https://go.awu.fyi/payment
